core/src/main/java/com/flipkart/gap/usl/core/processor/ExternalKafkaPublisher.java

```diff
@@ -0,0 +1,180 @@
+    }
+
+    public void process() throws ProcessingException {
+        try {
+            javaStreamingContext = getStreamingContext(applicationConfiguration.getEventProcessorConfig());
+            javaStreamingContext.start();
+            javaStreamingContext.awaitTermination();
+        } catch (Throwable e) {
+            throw new ProcessingException(e);
+        }
+    }
+
+    private JavaStreamingContext getStreamingContext(EventProcessorConfig eventProcessorConfig) throws Exception {
+
+        JavaStreamingContext javaStreamingContext = new JavaStreamingContext(sparkConf, Durations.seconds(eventProcessorConfig.getBatchDurationInSeconds()));
+
+        Map<TopicPartition, Long> topicPartitionMap = offsetManager.getMultipleTopicPartitions();
+        log.info("Fetched topic and partition map as {}", topicPartitionMap);
+        List<TopicPartition> topicPartitionList = new ArrayList<>(topicPartitionMap.keySet());
+        JavaInputDStream<ConsumerRecord<byte[], byte[]>> messages = KafkaUtils.createDirectStream(
+                javaStreamingContext,
+                LocationStrategies.PreferConsistent(),
+                ConsumerStrategies.Assign(topicPartitionList, kafkaParams, topicPartitionMap)
+        );
+
+        /*
+          This is to track offset ranges to persist it later in zookeeper.
+         */
+
+        final AtomicReference<OffsetRange[]> offsetRanges = new AtomicReference<>();
+        /*
+
+          Store offsets right after getting kafka stream.
+          Transform input events to dimension update events, create a tuple of composite key (entityId, dimensionName) and dimensionUpdate Event.
+
+         */
+        log.info("Starting transform to save offsets ");
+
+
+        messages.foreachRDD((VoidFunction2<JavaRDD<ConsumerRecord<byte[], byte[]>>, Time>) (consumerRecordJavaRDD, v2) -> {
+            OffsetRange[] offsets = ((HasOffsetRanges) consumerRecordJavaRDD.rdd()).offsetRanges();
+            offsetRanges.set(offsets);
+            for (OffsetRange offsetRange : offsets) {
+                log.info("Started Batch processing with offsets {},{},{},{}", offsetRange.topic(), offsetRange.partition(), offsetRange.fromOffset(), offsetRange.untilOffset());
+            }
+        });
+
+        messages.foreachRDD(
+                (VoidFunction2<JavaRDD<ConsumerRecord<byte[], byte[]>>, Time>) (consumerRecordJavaRDD, time) -> {
+
+                    JavaRDD<KafkaProducerRecord> publishedRDD = consumerRecordJavaRDD.map(consumerRecord -> {
+                        SparkHelper.bootstrap();
+                        return new KafkaProducerRecord(externalKafkaConfig.getTopicName(), consumerRecord.value());
+
+                    });
+
+                    try {
+                        KafkaPublisherDao kafkaPublisherDao = ExternalKafkaConfigurationModule.getInjector(applicationConfiguration).getInstance(KafkaPublisherDao.class);
+                        kafkaPublisherDao.sendRecords(publishedRDD.collect());
```

Comment:
this will send full partition record at once, better to split it into multiple batches and send in parallel